Do not fail node if SAML HTTP metadata is unavailable

tvernum opened the pull request:
This commit changes the SAML realm to use placeholder metadata (`UnresolvedEntity`) when the real metadata cannot be loaded over HTTPS - unless `metadata.http.fail_on_error` is set to true.

All future use of the realm will fail until the metadata is available, but this change allows the node to bootstrap successfully.

Closes: #37608

tvernum committed:
[WIP] Allow leniency in SAML HTTP metadata loading

This commit changes the SAML realm to use placeholder metadata
(`UnresolvedEntity`) when the real metadata cannot be loaded over
HTTPS - if `metadata.http.lenient` is set to true.

All future use of the realm will fail until the metadata is available,
but this change allows the node to bootstrap successfully.

Relates: #37608

elasticsearchmachine commented:
Hi @tvernum, I've created a changelog YAML for you.

tvernum committed:
Update docs/changelog/92810.yaml

tvernum committed:
Line length

jakelandis commented:
+1 to high level change to allow nodes to start if they can not download the metadata.

I haven't dug too deep so maybe this is already handled...but I do think we need a retry strategy such that if the HTTP issue is transient it will resolve itself without any intervention.  I think the 1 hour long default retry is too long in this case (and it is not obvious if the placeholder would get correctly updated), so we may want a hard coded shorter interval until the configured interval kicks in (i.e. with a default of 1 hour, retry every 1 minute during that hour). 

I don't see a need to expose this leniency behind a configuration option. If anything, I think the default should be reversed such that if set it will fail startup but only if it is configured to do so. Additionally, I think we should to emit a warn message if the http variant is used and the config is not set to fail fast to discourage usage of the http variant (mostly for trying to inform users setting this up for the first time to use a file). 

Also, is persisting this metadata in the .security index plausible/worth it such that we could use that as a fallback ? seemingly random things that break on restarts are the worst !

tvernum commented:
> I haven't dug too deep so maybe this is already handled...but I do think we need a retry strategy such that if the HTTP issue is transient it will resolve itself without any intervention. I think the 1 hour long default retry is too long in this case (and it is not obvious if the placeholder would get correctly updated), so we may want a hard coded shorter interval until the configured interval kicks in (i.e. with a default of 1 hour, retry every 1 minute during that hour)

I agree. I haven't tested whether OpenSAML does what we want yet, but it might. If not we will need to configure the refresh internal lower until we get the first successful metadata.

> I don't see a need to expose this leniency behind a configuration option

I think I agree, but it's hard to know what people really want.
The main downside of leniency is if you misconfigure the URL then the node starts up OK but SAML doesn't work and it might not be obvious why that is - although the volume of log messages that are generated should be an indication.
I've come to the conclusion that we probably want to default to leniency (despite it being abhorrent).

> Also, is persisting this metadata in the .security index plausible/worth it such that we could use that as a fallback ? 

Probably.
It's got the pain of having to coordinate between nodes & index health changes, but it should be possible - the metadata isn't huge.
I probably wouldn't try and do it in this PR because it would probably mean we can't ship as quickly, but I think it is possible to make it work.

tvernum committed:
Merge branch 'main' into saml-metadata-leniency

tvernum committed:
Auto reload failed SAML metadata

tvernum committed:
Merge branch 'main' into saml-metadata-leniency

tvernum committed:
Improve testing

tvernum committed:
Merge branch 'main' into saml-metadata-leniency

tvernum committed:
Merge branch 'main' into saml-metadata-leniency

tvernum committed:
Minor tweaks

tvernum commented:
This is ready for review now.

Mostly it's self explanatory, but there's a few bits that could use some additional details:
- In order to make this change as small as possible, I added a new `UnresolvedEntity` that is used as a placeholder whenever we can't load a real entity from the metadata URL. We could just use `null` or introduced an `Optional` but then we need to review all the places where the entity is used and make them handle null (or `empty`). The `UnresolvedEntity` throws exceptions when you try to use it so anything that tries to use the realm will fail at runtime, (hopefully) with a reasonable message.
- If the initial load of metadata fails, we want to keep trying (more frequently than the default 1h refresh). That was solved through two complimentary ways:
   1. I added a new `metadata.http.minimum_refresh` setting, which defaults to 5 minutes. The way it's used in OpenSAML changes the semantics of the realm slightly, but mostly in helpful ways. If metadata has not been loaded, it will be retried every 5 minutes. If metadata has been loaded, and the URL has an `Expires` header that is less than 1 hour in the future, then it will be refreshed when it expires (with a minimum of 5 minutes)
   2. When we request the metadata from the resolver, if it returns `null` (which means we never received valid metadata for this entity) we automatically force a refresh (once). An `AtomicBoolean` ensures that only 1 thread per node will refresh at a time, the others will fail with "no metadata" as they always have.

I still have 2 more tasks to do, but I'd like to go through a first round of reviews first:
1. Update docs
2. Write for the case where there are 2 SAML realms - 1 has valid metadata, the other is in a failed state. We need to ensure that authentication with the valid realm works even if the other realm is broken.

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

elasticsearchmachine commented:
Hi @tvernum, I've updated the changelog YAML for you.

tvernum committed:
Update docs/changelog/92810.yaml

tvernum committed:
Avoid negative delay bug in OpenSAML

jakelandis reviewed (commented):
Change are looking good. I like the placeholder and refresh strategy. A couple minor questions and will review the tests when complete.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/saml/SamlRealmSettings.java:
```diff
@@ -47,6 +47,18 @@ public class SamlRealmSettings {
+    public static final Setting.AffixSetting<TimeValue> IDP_METADATA_HTTP_MIN_REFRESH = Setting.affixKeySetting(
+        RealmSettings.realmSettingPrefix(TYPE),
+        IDP_METADATA_SETTING_PREFIX + "http.minimum_refresh",
+        key -> Setting.timeSetting(key, TimeValue.timeValueMinutes(5), Setting.Property.NodeScope)
```
> should we also set a min value for this setting to prevent absurdly low values

We can. I'm not sure what would constitute absurdly low though. I guess "minutes" is sensible, "seconds" is acceptable and "milliseconds" is a problem. I'm not sure what the precise dividing line would be (my inclination is 5s)

> maybe we should we add setting validation that this value is < http.refresh rather than setting them equal at runtime, i.e. prefer strict validation vs. lenient behavior

I considered that, but it means either a potential breaking change or a slightly more complex behaviour.
If an admin has set `http.refresh` to _4 minutes_ and `http.minimum_refresh` defaults to _5 minutes_ then the introduction of this new setting would prevent their node from starting.

To avoid that we need to only apply the validation if `http.minimum_refresh` has an explicit value, and keep the current behaviour otherwise.
  
> should this be dynamic?

It would be helpful if `http.refresh` was dynamic, because if you knew that you had published updated (remote) metadata, you could set the `refresh` time down to force it to be reloaded more quickly. 
It's not so important to set `http.minimum_refresh` dynamically because of the automatic reload behaviour you mention. 

> should we deprecate the http.refresh in favor of http.max_refresh?

We could. It's arguably more consistent, but my gut is that it's one of the annoying changes that isn't worth it.

I don't think having `http.refresh` and `http.minimum_refresh` is terribly confusing, even if it isn't consistent - in fact I expect most admins who need to fiddle with refresh times should just set  `http.refresh` and ignore `http.minimum_refresh` anyway.

On balance, forcing every admin who has configured `http.refresh` to go through the deprecation process doesn't seem justified just to make things a bit more consistent - it doesn't feel like it meets the necessary threshold to force that pain onto users.

tvernum replied:
> I'm not sure what the precise dividing line would be (my inclination is 5s)

I ended up setting the minimum to 500ms. Mostly because I had a test that relied on setting it to a very low value and making sure it did a refresh in the background. If the limit was 5s then the test would take >5s to run (for no great reason).

500ms still stops people setting it to tiny values, and if they really want it to refresh twice every second, I guess they can.

tvernum committed:
Better validation around minimum_refresh

ywangd commented:
Could you please help me understand the following behaviour?

> If metadata has been loaded, and the URL has an Expires header that is less than 1 hour in the future, then it will be refreshed when it expires (with a minimum of 5 minutes)

What if the refresh at expiration time fails? Does OpenSAML fallback to refresh every 5 minutes? Also, what will be the return value of `resolveSingle()`? Is it `null`?

tvernum commented:
> What if the refresh at expiration time fails? Does OpenSAML fallback to refresh every 5 minutes? Also, what will be the return value of resolveSingle()? Is it null?

The underlying code in OpenSAML is complex with many separate execution paths. My best understanding is:

1. If there is any failure to refresh the metadata, we will continue to use the old metadata (if we successfully loaded metadata at some point). The exception would be if the "failure" is actually a `200` response that parses successfully as XML metadata (which is highly implausible).
2. The min/max `refresh` times just control how frequently OpenSAML will attempt to refresh. It will attempt to refresh at least every `maxRefresh` time-units, and never attempt to refresh more often than `minRefresh` time-units. Within those boundaries it attempts to pick the best "nextScheduledRefresh" based whether it ever received a successful HTTP response, when that was, what `Expires` header it received (I believe) and `validUntil` (etc) attributes inside the metadata itself.

It should never cause us any problems - once we've loaded metadata we never lose it, though it could expire while unsuccessfully attempting to load new metadata.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlRealm.java:
```diff
@@ -676,10 +691,29 @@ private static Tuple<AbstractReloadingMetadataResolver, Supplier<EntityDescripto
+        TimeValue maxRefresh = config.getSetting(IDP_METADATA_HTTP_REFRESH);
+        TimeValue minRefresh = config.getSetting(IDP_METADATA_HTTP_MIN_REFRESH);
+        if (minRefresh.compareTo(maxRefresh) > 0) {
+            if (config.hasSetting(IDP_METADATA_HTTP_MIN_REFRESH)) {
+                throw new SettingsException(
+                    "the value ({}) for [{}] cannot be greater than the value ({}) for [{}]",
+                    minRefresh.getStringRep(),
+                    RealmSettings.getFullSettingKey(config, IDP_METADATA_HTTP_MIN_REFRESH),
+                    maxRefresh.getStringRep(),
+                    RealmSettings.getFullSettingKey(config, IDP_METADATA_HTTP_REFRESH)
+                );
```
Nit: It is also possible that user explicitly configures the max refresh to be lower than the default min refresh. In that case, this check won't be able to catch it.

tvernum replied:
Yes, that's intentional for BWC.
If someone has an existing config with `.refresh` set to 4 minutes, then we don't want the node to fail during bootstrap, even though it their refresh interval conflicts with the default for `.minimum_refresh`

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlRealm.java:
```diff
@@ -744,25 +785,69 @@ private static Tuple<AbstractReloadingMetadataResolver, Supplier<EntityDescripto
+                    throw SamlUtils.samlException("Cannot find metadata for entity [{}] in [{}]", entityId, sourceLocation);
+                } else {
+                    logger.warn(
+                        "cannot load SAML metadata for [{}] from [{}]; SAML authentication for this realm will fail",
```
Nit: might want to include realm name here. It can be derived by the entityId and url. But it is convenient for readers to just call it out.

tvernum replied:
Unfortunately, we don't know the realm name here and it's tricky to resolve that.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlRealm.java:
```diff
@@ -744,25 +785,69 @@ private static Tuple<AbstractReloadingMetadataResolver, Supplier<EntityDescripto
+                    REFRESHING_METADATA.compareAndSet(true, false);
+                }
+            }
+            descriptor = resolver.resolveSingle(criteria);
```
Is it intentional that `resolver.resolveSingle` is called again even when the above `if` check returns false, i.e. when the thread is not performing the actual refresh?

tvernum replied:
It is intentional, on the off chance that whichever other thread performed the refresh has completed in time. It will likely fail, but sometimes it will work.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlRealm.java:
```diff
@@ -744,25 +785,69 @@ private static Tuple<AbstractReloadingMetadataResolver, Supplier<EntityDescripto
+                } catch (ResolverException e) {
+                    // Refresh failed. Since there could be a lot of these and it's only a saving mechanism we log as debug
+                    // The resolver itself will log a WARN with the HTTP status etc.
+                    logger.debug(() -> "Failed to refresh SAML metadata for [" + entityId + "]", e);
```
Should we have a corresponding log for when refresh is successful?

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlRealm.java:
```diff
@@ -744,25 +785,69 @@ private static Tuple<AbstractReloadingMetadataResolver, Supplier<EntityDescripto
+             * 1. We didn't find any metadata file (e.g. the metadata is loaded over http, and the request returned an error
+             * 2. The file does exist, but doesn't contain the entity.
+             * In either case it's worth refreshing the metadata again to see if it's now correct because it's possible that the problem
+             *    has been resolved (although if metadata is loaded from a local file we monitor it for changes anyway, so this refresh
+             *    is unlikely to have any benefit).
```
Can we check the `resolver` object type and only perform refresh if it is a `HTTPMetadataResolver`?

tvernum replied:
We can, do we care enough about that to make the code more complex?

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/UnresolvedEntity.java:
```diff
@@ -0,0 +1,409 @@
+    }
+
+    @Override
+    public void setExtensions(Extensions extensions) {
+
+    }
```
UnsupportedOperationException?

ywangd commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlRealmTests.java:
```diff
@@ -185,6 +172,177 @@ public void testReadIdpMetadataFromHttps() throws Exception {
+            final TimeValue minimumRefreshTime = testBackgroundRefresh ? TimeValue.timeValueMillis(500) : defaultRefreshTime;
+
+            final Tuple<RealmConfig, SSLService> config = buildConfig("https://localhost:" + webServer.getPort(), builder -> {
+                if (defaultRefreshTime != null) {
```
I don't think this null check is necessary since it's never null.

ywangd commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlRealmTests.java:
```diff
@@ -185,6 +172,177 @@ public void testReadIdpMetadataFromHttps() throws Exception {
+                        defaultRefreshTime.getStringRep()
+                    );
+                }
+                if (minimumRefreshTime != null) {
```
Similarly, null check is not needed.

ywangd reviewed (approved):
LGTM

I left some comments, but do not need to look at the changes again. Thanks!

tvernum committed:
Merge branch 'main' into saml-metadata-leniency

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlRealm.java:
```diff
@@ -155,6 +158,8 @@ public final class SamlRealm extends Realm implements Releasable {
     // Although we only use this for IDP metadata loading, the SSLServer only loads configurations where "ssl." is a top-level element
     // in the realm group configuration, so it has to have this name.
 
+    private static final AtomicBoolean REFRESHING_METADATA = new AtomicBoolean(false);
```
I'd love to make it an instance level field, but it's a messy refactoring - all the code that loads metadata is static.
It would be good to clean that up (I think moving the metadata resolution into a separate class would be ideal) but I don't think it can be tackled in this PR.
Let me know if you feel strongly that it needs to change.

ywangd replied:
I'd be happy with just a comment explaining the choice here. Thanks!

tvernum committed:
Address feedback

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/saml/SamlRealmSettings.java:
```diff
@@ -47,6 +47,18 @@ public class SamlRealmSettings {
+
+    public static final Setting.AffixSetting<Boolean> IDP_METADATA_HTTP_FAIL_ON_ERROR = Setting.affixKeySetting(
+        RealmSettings.realmSettingPrefix(TYPE),
+        IDP_METADATA_SETTING_PREFIX + "http.fail_on_error",
+        key -> Setting.boolSetting(key, false, Setting.Property.NodeScope)
```
Yeah, @jakelandis and I discussed earlier. My first version reversed the default behavior and would fail unless you made it lenient. We agreed that it was better to change the behavior with an option to revert and that we didn't consider it breaking.

tvernum commented:
I'm going to move ahead with docs + a QA test on this since it seems like we have consensus on the approach.

elasticsearchmachine commented:
Hi @tvernum, I've updated the changelog YAML for you.

tvernum committed:
Update docs/changelog/92810.yaml

tvernum committed:
Add QA Test for SAML metadata leniency

tvernum committed:
Merge branch 'main' into saml-metadata-leniency

tvernum committed:
Convert mock webserver to ClassRule

tvernum committed:
Document new settings

tvernum commented:
@elasticmachine run elasticsearch-ci/part-3 please

tvernum commented:
@jakelandis, @ywangd I've added a QA test & docs for the new settings. Do either of you want to review those?

ywangd commented on docs/reference/settings/security-settings.asciidoc:
```diff
@@ -1089,21 +1089,62 @@ capabilities and configuration of the Identity Provider.
+`idp.metadata.http.fail_on_error`::
+(<<static-cluster-setting,Static>>)
+If set to `true`, the realm will fail on startup (and prevent the node from starting)
+if it attempts to load metadata over HTTPs and that metadata is not available.
```
Nit: It says `HTTPs` here, but it was `HTTPS` in the above.

ywangd commented on x-pack/plugin/security/qa/saml-rest-tests/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/saml/SamlServiceProviderMetadataIT.java:
```diff
@@ -0,0 +1,254 @@
+    }
+
+    public void testAuthenticationWhenMetadataIsUnreliable() throws Exception {
+        // Start with no metadata available
+        makeAllMetadataUnavailable();
```
Is this necessary? The httpServer is configured to _not_ return metadata in the setup. If that is not the case, it will be too late here to mark them available because the ES cluster is already started. So I don't think this does much. Maybe we can replace it with an assertion that all metadata is indeed unavailable.

ywangd commented:
I had a brief look at the doc and test update. They look good to me. The way you setup the tests is quite intriguing. It will be a great reference when I need to juggle so much of SAML messages :)

tvernum committed:
Address feedback

tvernum commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into saml-metadata-leniency

jakelandis commented on test/framework/src/main/java/org/elasticsearch/test/junit/RunnableTestRuleAdapter.java:
```diff
@@ -0,0 +1,34 @@
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+public class RunnableTestRuleAdapter implements TestRule {
```
nit: can you add some java doc with why this exists and an example usage ?

jakelandis reviewed (approved):
LGTM (and really nice work on the integration test and setup!)

tvernum committed:
Merge branch 'main' into saml-metadata-leniency

tvernum committed:
Add javadoc
